librz/core/creg.c

```diff
@@ -0,0 +1,96 @@
+// SPDX-FileCopyrightText: 2021 Florian Märkl <info@florianmaerkl.de>
+// SPDX-FileCopyrightText: 2021 ret2libc <sirmy15@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_core.h>
+
+/**
+ * \brief Get the currently relevant RzReg
+ *
+ * Depending on whether or not the core is in debug mode, this returns the RzReg
+ * from debug or analysis (emulation).
+ * Before using this function, think twice whether it would not make more sense
+ * to use core->dbg->reg or analysis->dbg->reg directly, depending on what you want
+ * to do with it.
+ */
+RZ_API RzReg *rz_core_reg_default(RzCore *core) {
+	return rz_core_is_debug(core) ? core->dbg->reg : core->analysis->reg;
+}
+
+/// rz_reg_getv_by_role_or_name() on rz_core_reg_default()
+RZ_API ut64 rz_core_reg_getv_by_role_or_name(RzCore *core, const char *name) {
+	// this logic has to be in sync with rz_core_reg_default().
+	if (rz_core_is_debug(core)) {
+		// call this instead of rz_reg_getv_... directly because it also syncs
+		return rz_debug_reg_get(core->dbg, name);
+	}
+	return rz_reg_getv_by_role_or_name(core->analysis->reg, name);
+}
+
+/// set on rz_core_reg_default()
+RZ_API bool rz_core_reg_set_by_role_or_name(RzCore *core, const char *name, ut64 num) {
+	// this logic has to be in sync with rz_core_reg_default().
+	if (rz_core_is_debug(core)) {
+		// call this instead of rz_reg_getv_... directly because it also syncs
+		return rz_debug_reg_set(core->dbg, name, num);
+	}
+	RzRegItem *ri = rz_reg_get_by_role_or_name(core->analysis->reg, name);
+	if (!ri) {
+		return false;
+	}
+	return rz_reg_set_value(core->analysis->reg, ri, num);
```

Comment:
shouldn't this become:
```C
reg = rz_core_reg_default(core);
RzRegItem *ri = rz_reg_get_by_role_or_name(reg, name);
return rz_reg_set_value(reg, ri, num);
```
?

Reply:
It is ok like this because `rz_core_is_debug(core)` has already been checked. This is what the comment `this logic has to be in sync with rz_core_reg_default()` means.

Reply:
Yes but why do you want to sync regs in one case and not in the other?

Reply:
It's only for syncing with the debuggee. In the case of analysis->regs, there is no such thing.

Reply:
Don't you need to sync the registers with the flags anyway?

Reply:
Theoretically yes, but flags are completely separate, it's done by the caller and not here. In general I am not a fan at all of all the sync calls for both debug and flags. At least flags could be done much better with events, but that is another issue.

Reply:
I'm just not sure about this because callers of this function would have different visible results based on debug vs non-debug mode. In debug mode they can expect flags to be synced, in non-debug mode they can't. That means that the caller needs to be aware of what this function does, which counters a bit the purpose of having a unified function IMO.

Reply:
No, flags are not synced at all in this function, regardless of the mode.

Reply:
Oh, I thought `because it also syncs` was referring to the flags, sorry

Reply:
After looking at more code that will use this, I have changed my mind about syncing flags. It now also syncs the flags and that happens in both debug and emulation mode.